Fix TaskInstance actions with upstream/downstream

ashb opened the pull request:
When we added clearing individual mapped tasks, we unfortunately broke
the up/down stream feature

This was because when passing task_id/task_id+map_index down that
limited it to _just_ that task_id.

So we need to change it to also support the up/downstream we need to add
those tasks to the list we pass on, meaning that we have to support
task_id and (task_id,map_index) tuples in the same `task_id` list.

This almost certainly needs tests adding!
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragement file, named `{pr_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

ashb committed:
Fix Clear+upstream/downstream

When we added clearing individual mapped tasks, we unfortunately broke
the up/down stream feature

This was because when passing task_id/task_id+map_index down that
limited it to _just_ that task_id.

So we need to change it to also support the up/downstream we need to add
those tasks to the list we pass on, meaning that we have to support
task_id and (task_id,map_index) tuples in the same `task_id` list.

jedcunningham commented on airflow/models/dag.py:
```diff
@@ -1439,19 +1439,34 @@ def _get_task_instances(
             tis = tis.filter(DagRun.execution_date >= start_date)
 
+        def _task_id_map_filter(val):
+            # Compute a filter  for TI.task_id and TI.map_index based on input values
```
```suggestion
            # Compute a filter for TI.task_id and TI.map_index based on input values
```

ashb committed:
Fix Mark Success/Failure

Similar problems as clear.

Note we changed the eager loading of DagRun (which is also the default)
in to an explicit lazy load to avoid a needless join on to DR for this
function.

uranusjr commented on airflow/models/taskinstance.py:
```diff
@@ -228,14 +229,14 @@ def clear_task_instances(
             session.merge(ti)
 
-        task_id_by_key[ti.dag_id][ti.run_id][ti.try_number].add(ti.task_id)
+        task_id_by_key[ti.dag_id][ti.run_id][ti.map_index][ti.try_number].add(ti.task_id)
```
This mapping is _so_ weird.

uranusjr commented on airflow/models/taskinstance.py:
```diff
@@ -2529,6 +2536,28 @@ def filter_for_tis(tis: Iterable[Union["TaskInstance", TaskInstanceKey]]) -> Opt
+            filters.append(
+                tuple_in_condition((cls.task_id, cls.map_index), with_map_index),
+            )
+        return or_(*filters) if len(filters) > 1 else filters[0]
```
```suggestion
        if not filters:
            return true()
        elif len(filters) == 1:
            return filters[0]
        return or_(*filters)
```

Calling `or_()` without any arguments is deprecated https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.or_

ashb replied:
In my head this function should never result in 0 filters. I should assert that

ashb committed:
Keep functions public

ashb committed:
Update tests for set_state API fix

ashb committed:
Tests for dag.clear change

ashb committed:
Code review

uranusjr reviewed (approved):
Only part I can’t make sense of is the dag_id-run_id-try_number-map_index lookup. I’ll take your word they are correct.

uranusjr committed:
Improve test debugging display

ashb committed:
Fix tests - join, not cartesian product

ashb committed:
Fix mark subdag test

The comment doesn't make anysense, using a snapshot just works fine.

uranusjr committed:
Avoid checking list with unstable order
